ops/pebble.py

```diff
@@ -1926,3 +1910,188 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+
+
+class MultipartLargeFileParser:
+    """A limited purpose multi-part parser backed by files for memory efficiency."""
+
+    def __init__(self, boundary: bytes):
+        self._buffer = b''
+        self._response = None
+        self._files = {}
+
+        # Store the initial FSM state.  See _run_parser_fsm() for a full explanation.
+        self._next_state_fn = self._state_detect_boundary
+
+        # Prepare the MIME multipart boundary line matcher.
+        # Note: RFC 2046 notes optional "linear whitespace" (e.g. [ \t]) after the boundary pattern
+        # and the optional "--" suffix.  Pebble doesn't use this, so to simplify implementation I
+        # am ignoring it.
+        boundary_rgx_pattern = r"--{}(--)?\r\n".format(boundary)
+        self._boundary_rgx = re.compile(boundary_rgx_pattern.encode())
+
+        # When streaming file data to tempfiles, we need to make sure we don't feed part of the
+        # boundary line by mistake.  This is easy if we detect the boundary in full, but if we
+        # haven't received it yet, we need to be careful not to feed any partial boundary line
+        # to disk.  This minimum buffer length threshold will ensure that, and will work for both
+        # regular and terminating boundary cases.
+        #
+        # Calculated as: boundary header length + prefix/suffix -- (4 bytes) + \r\n (2 bytes)
+        self._safe_buffer_len = len(boundary) + 6
+
+    def __del__(self):
+        for file in self._files.values():
+            os.unlink(file.name)
+
+    def feed(self, data: bytes):
+        """Provide more data to the parser to process."""
+        self._buffer += data
+        self._run_parser_fsm()
+
+    def _run_parser_fsm(self):
+        """Runs the parser FSM.
+
+        Each FSM state is represented by a function.  Transitions to other FSM states are done by
+        returning the next state's function.  If None is returned, the FSM will be terminated.
+
+        As a special case, to avoid the need to transfer to and from an explicit paused state when
+        we run out of pending data to parse, we use the StopParsing exception to signal that the
+        FSM should be paused.  It will be restarted from the same state the next time data is
+        received.
+        """
+        try:
+            while self._next_state_fn:
+                self._next_state_fn = self._next_state_fn()
+        except StopParsing:
+            pass
+
+    def _state_detect_boundary(self):
+        match = self._boundary_rgx.search(self._buffer)
+        if match is None:
+            # Couldn't find a complete boundary in the buffer; stop for now
+            raise StopIteration()
+
+        # This is here as a paranoia check; we should only be in this state when the buffer
+        # immediately starts with the boundary line.
+        span = match.span()
+        data_before = self._buffer[:span[0]]
+        if data_before:
+            raise RuntimeError(
```

Comment:
Would this (and similar ones below) be better as a `ProtocolError`? Because it's invalid data coming from Pebble, right?

Reply:
Ping on this.